File: src/simplemd.f90
!!!   This file is part of simplemd
!!!
!!!   simplemd is free software: you can redistribute it and/or modify
!!!   it under the terms of the GNU General Public License as published by
!!!   the Free Software Foundation, either version 3 of the License, or
!!!   (at your option) any later version.
!!!
!!!   Foobar is distributed in the hope that it will be useful,
!!!   but WITHOUT ANY WARRANTY; without even the implied warranty of
!!!   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!!!   GNU General Public License for more details.
!!!
!!!   You should have received a copy of the GNU General Public License
!!!   along with Foobar.  If not, see <http://www.gnu.org/licenses/>.

program simplemd
use routines
implicit none

integer              :: natoms          ! number of atoms
real,    allocatable :: positions(:,:)  ! atomic positions
real,    allocatable :: velocities(:,:) ! velocities
                                        ! was calculated last time
real,    allocatable :: masses(:)       ! masses
real,    allocatable :: forces(:,:)     ! forces   
real                 :: cell(3)         ! cell size

! neighbour list variables
! see Allen and Tildesey book for details
integer              :: listsize        ! size of the list array
integer, allocatable :: list(:)         ! neighbour list
integer, allocatable :: point(:)        ! pointer to neighbour list
real,    allocatable :: positions0(:,:) ! reference atomic positions, i.e. positions when the neighbour list

! input parameters
! all of them have a reasonable default value, set in read_input()
real           :: tstep          ! simulation timestep
real           :: temperature    ! temperature
real           :: friction       ! friction for Langevin dynamics (for NVE, use 0)
real           :: listcutoff     ! cutoff for neighbour list
real           :: forcecutoff    ! cutoff for forces
integer        :: nstep          ! number of steps
integer        :: nconfig        ! stride for output of configurations
integer        :: nstat          ! stride for output of statistics
integer        :: maxneighbour   ! maximum average number of neighbours per atom
integer        :: idum           ! seed
logical        :: wrapatoms      ! if true, atomic coordinates are written wrapped in minimal cell
character(256) :: inputfile      ! name of file with starting configuration (xyz)
character(256) :: outputfile     ! name of file with final configuration (xyz)
character(256) :: trajfile       ! name of the trajectory file (xyz)
character(256) :: statfile       ! name of the file with statistics


real    :: engkin         ! kinetic energy
real    :: engconf        ! configurational energy
real    :: engint         ! integral for conserved energy in Langevin dynamics

logical :: recompute_list ! control if the neighbour list have to be recomputed
integer :: istep          ! step counter
integer :: iatom

call read_input(temperature,tstep,friction,forcecutoff, &
                listcutoff,nstep,nconfig,nstat, &
                wrapatoms, &
                inputfile,outputfile,trajfile,statfile, &
                maxneighbour,idum)

! number of atoms is read from file inputfile
call read_natoms(inputfile,natoms)

! write the parameters in output so they can be checked
write(*,*) "Starting configuration           : ",trim(inputfile)
write(*,*) "Final configuration              : ",trim(outputfile)
write(*,*) "Number of atoms                  : ",natoms
write(*,*) "Temperature                      : ",temperature
write(*,*) "Time step                        : ",tstep
write(*,*) "Friction                         : ",friction
write(*,*) "Cutoff for forces                : ",forcecutoff
write(*,*) "Cutoff for neighbour list        : ",listcutoff
write(*,*) "Number of steps                  : ",nstep
write(*,*) "Stride for trajectory            : ",nconfig
write(*,*) "Trajectory file                  : ",trim(trajfile)
write(*,*) "Stride for statistics            : ",nstat
write(*,*) "Statistics file                  : ",trim(statfile)
write(*,*) "Max average number of neighbours : ",maxneighbour
write(*,*) "Seed                             : ",idum
write(*,*) "Are atoms wrapped on output?     : ",wrapatoms

! Since each atom pair is counted once, the total number of pairs
! will be half of the number of neighbours times the number of atoms
listsize=maxneighbour*natoms/2

! allocation of dynamical arrays
allocate(positions(3,natoms))
allocate(positions0(3,natoms))
allocate(velocities(3,natoms))
allocate(forces(3,natoms))
allocate(masses(natoms))
allocate(point(natoms))
allocate(list(listsize))

! masses are hard-coded to 1
masses=1.0

! energy integral initialized to 0
engint=0.0

! positions are read from file inputfile
call read_positions(inputfile,natoms,positions,cell)

! velocities are randomized according to temperature
call randomize_velocities(natoms,temperature,masses,velocities,idum)

! neighbour list are computed, and reference positions are saved
call compute_list(natoms,listsize,positions,cell,listcutoff,point,list)
write(*,*) "List size: ",point(natoms)-1
positions0=positions

! forces are computed before starting md
call compute_forces(natoms,listsize,positions,cell,forcecutoff,point,list,forces,engconf)

! here is the main md loop
! Langevin thermostat is applied before and after a velocity-Verlet integrator
! the overall structure is:
!   thermostat
!   update velocities
!   update positions
!   (eventually recompute neighbour list)
!   compute forces
!   update velocities
!   thermostat
!   (eventually dump output informations)
do istep=1,nstep
  call thermostat(natoms,masses,0.5*tstep,friction,temperature,velocities,engint,idum)

  do iatom=1,natoms
    velocities(:,iatom)=velocities(:,iatom)+forces(:,iatom)*0.5*tstep/masses(iatom)
  end do

  do iatom=1,natoms
    positions(:,iatom)=positions(:,iatom)+velocities(:,iatom)*tstep
  end do

! a check is performed to decide whether to recalculate the neighbour list
  call check_list(natoms,positions,positions0,listcutoff,forcecutoff,recompute_list)
  if(recompute_list) then
    call compute_list(natoms,listsize,positions,cell,listcutoff,point,list)
    positions0=positions
    write(*,*) "Neighbour list recomputed at step ",istep
    write(*,*) "List size: ",point(natoms)-1
  end if

  call compute_forces(natoms,listsize,positions,cell,forcecutoff,point,list,forces,engconf)

  do iatom=1,natoms
    velocities(:,iatom)=velocities(:,iatom)+forces(:,iatom)*0.5*tstep/masses(iatom)
  end do

  call thermostat(natoms,masses,0.5*tstep,friction,temperature,velocities,engint,idum)

! kinetic energy is calculated
  call compute_engkin(natoms,masses,velocities,engkin)

! eventually, write positions and statistics
  if(modulo(istep,nconfig)==0) call write_positions(trajfile,natoms,positions,cell,wrapatoms)
  if(modulo(istep,nstat)==0)   call write_statistics(statfile,istep,tstep,natoms,engkin,engconf,engint)

end do

call write_final_positions(outputfile,natoms,positions,cell,wrapatoms)
write(*,*) "Execution completed"

! deallocation of all allocatable array
deallocate(positions)
deallocate(velocities)
deallocate(forces)
deallocate(masses)
deallocate(positions0)
deallocate(point)
deallocate(list)

end program simplemd

File: python/simplemd.py
import numpy as np
import re
import numba
import tempfile
import sys

try:
    from tqdm import tqdm
except ModuleNotFoundError:
    def tqdm(a):
        return a

# Random number generator compatible with C++ and FORTRAN versions

_IA=16807
_IM=2147483647
_IQ=127773
_IR=2836
_NTAB=32
_NDIV=(1+(_IM-1)//_NTAB)
_EPS=3.0e-16
_AM=1.0/_IM
_RNMX=(1.0-_EPS)

@numba.njit(cache=True,fastmath=True)
def _U01(idum,iy,iv):
    if idum <= 0 or iy == 0:
        if (-idum < 1):
            idum=1
        else:
            idum=-idum
        for j in range(_NTAB+7,-1,-1):
            k=idum//_IQ
            idum=_IA*(idum-k*_IQ)-_IR*k
            if idum<0:
                idum += _IM
            if j < _NTAB:
                iv[j] = idum
        iy=iv[0]
    k=idum//_IQ
    idum=_IA*(idum-k*_IQ)-_IR*k
    if idum<0:
        idum += _IM
    j=iy//_NDIV
    iy=iv[j]
    iv[j]=idum
    temp=_AM*iy
    if temp > _RNMX:
        temp = _RNMX
    return idum,iy,temp

@numba.njit(cache=True,fastmath=True)
def _U01_loop(idum,iy,iv,numbers):
    for i in range(len(numbers)):
        idum,iy,numbers[i] = _U01(idum,iy,iv)
    return idum,iy

@numba.njit(cache=True,fastmath=True)
def _Gaussian(idum,iy,iv,switchGaussian,saveGaussian):
    if(switchGaussian):
        return idum,iy,False,0.0,saveGaussian
    else:
        while True:
            idum,iy,r1=_U01(idum,iy,iv)
            idum,iy,r2=_U01(idum,iy,iv)
            v1=2.0*r1-1.0
            v2=2.0*r2-1.0
            rsq=v1*v1+v2*v2
            if(rsq<1.0 and rsq>0.0):
                break
        fac=np.sqrt(-2.0*np.log(rsq)/rsq)
        return idum,iy,True,v1*fac,v2*fac

@numba.njit(cache=True,fastmath=True)
def _Gaussian_loop(idum,iy,iv,switchGaussian,saveGaussian,numbers):
    for i in range(len(numbers)):
        idum,iy,switchGaussian,saveGaussian,numbers[i] = _Gaussian(
            idum,iy,iv,switchGaussian,saveGaussian
    )
    return idum,iy,switchGaussian,saveGaussian

class Random():
    def __init__(self,seed=0):
        self.switchGaussian=False
        self.saveGaussian=0.0
        self.iy=0
        self.iv=np.zeros(_NTAB, dtype=int)
        self.idum=0
        self.idum=seed
    def U01(self,shape=None):
        if shape is None:
            self.idum,self.iy,temp = _U01(self.idum,self.iy,self.iv)
            return temp
        else:
            numbers=np.zeros(np.prod(shape))
            self.idum,self.iy = _U01_loop(self.idum,self.iy,self.iv,numbers)
            return numbers.reshape(shape)

    def Gaussian(self,shape=None):
        if shape is None:
            self.idum,self.iy,self.switchGaussian,self.saveGaussian,temp=_Gaussian(
                self.idum,self.iy,self.iv,self.switchGaussian,self.saveGaussian
            )
            return temp
        else:
            numbers=np.zeros(np.prod(shape))
            self.idum,self.iy,self.switchGaussian,self.saveGaussian=_Gaussian_loop(
                self.idum,self.iy,self.iv,self.switchGaussian,self.saveGaussian,numbers
            )
            return numbers.reshape(shape)


@numba.njit(cache=True,fastmath=True)
def _compute_forces(cell, positions, forcecutoff, neighbors, point, forces):
    engconf=0.0
    forces.fill(0.0)
    forcecutoff2=forcecutoff*forcecutoff
    engcorrection=4.0*(1.0/forcecutoff2**6-1.0/forcecutoff2**3)

    for i in range(len(positions)):
        for j in range(point[i],point[i+1]):
            ja=neighbors[j]
            distancex=positions[i,0]-positions[ja,0]
            distancey=positions[i,1]-positions[ja,1]
            distancez=positions[i,2]-positions[ja,2]
            distancex-=np.floor(distancex/cell[0]+0.5)*cell[0]
            distancey-=np.floor(distancey/cell[1]+0.5)*cell[1]
            distancez-=np.floor(distancez/cell[2]+0.5)*cell[2]
            distance2=distancex**2+distancey**2+distancez**2
            if distance2 <= forcecutoff2:
                invdistance2=1.0/distance2
                invdistance6=invdistance2*invdistance2*invdistance2
                e=4.0*invdistance6*invdistance6-4.0*invdistance6-engcorrection
                engconf+=e
                fmod=2.0*4.0*(6.0*invdistance6*invdistance6-3.0*invdistance6)*invdistance2
                fx=fmod*distancex
                fy=fmod*distancey
                fz=fmod*distancez
                forces[i,0]+=fx
                forces[i,1]+=fy
                forces[i,2]+=fz
                forces[ja,0]-=fx
                forces[ja,1]-=fy
                forces[ja,2]-=fz
    return engconf

@numba.njit(cache=True,fastmath=True)
def _compute_list(cell,positions,listcutoff,nlist,point):
   listcutoff2=listcutoff**2
   point[0]=0
   for i in range(len(positions)):
       point[i+1]=point[i]
       for j in range(i+1,len(positions)):
            distancex=positions[i,0]-positions[j,0]
            distancey=positions[i,1]-positions[j,1]
            distancez=positions[i,2]-positions[j,2]
            distancex-=np.floor(distancex/cell[0]+0.5)*cell[0]
            distancey-=np.floor(distancey/cell[1]+0.5)*cell[1]
            distancez-=np.floor(distancez/cell[2]+0.5)*cell[2]
            distance2=distancex**2+distancey**2+distancez**2
            if distance2 <= listcutoff2:
                if point[i+1]>=len(nlist):
                    raise Exception("Verlet list size exceeded\nIncrease maxneighbours")
                nlist[point[i+1]]=j
                point[i+1]+=1


def generate_lattice(n=1, a0=1.6796):
    coord=np.zeros((4*n*n*n,3))
    l=0
    for i in range(n):
       for j in range(n):
           for k in range(n):
               coord[l]=(i,j,k)
               l+=1
               coord[l]=(i+0.5,j,k+0.5)
               l+=1
               coord[l]=(i+0.5,j+0.5,k)
               l+=1
               coord[l]=(i,j+0.5,k+0.5)
               l+=1
    return ((n*a0,n*a0,n*a0),coord*a0)

def read_input(file):
    keys={}
    with open(file,"r") as f:
        for line in f:
            line=re.sub("#.*$","",line)
            line=re.sub(" *$","",line)
            words=line.split()
            if len(words)==0:
                continue
            key=words[0]
            if key=="temperature":
                keys["temperature"]=float(words[1])
            elif key=="tstep":
                keys["tstep"]=float(words[1])
            elif key=="friction":
                keys["friction"]=float(words[1])
            elif key=="forcecutoff":
                keys["forcecutoff"]=float(words[1])
            elif key=="listcutoff":
                keys["listcutoff"]=float(words[1])
            elif key=="nstep":
                keys["nstep"]=int(words[1])
            elif key=="nconfig":
                keys["nconfig"]=int(words[1])
                keys["trajfile"]=words[2]
            elif key=="nstat":
                keys["nstat"]=int(words[1])
                keys["statfile"]=words[2]
            elif key=="wrapatoms":
                if re.match("[Tt].*",words[1]):
                    keys["wrapatoms"]=True
            elif key=="maxneighbours":
                keys["maxneighbors"]=int(words[1])
            elif key=="inputfile":
                keys["inputfile"]=words[1]
            elif key=="outputfile":
                keys["outputfile"]=words[1]
            elif key=="idum":
                keys["idum"]=int(words[1])
            else:
                raise Exception("Unknown keyword: "+key)
    return keys

def read_positions(file):
    with open(file,"r") as f:
        natoms=int(f.readline())
        cell=[float(x) for x in f.readline().split()]
        positions=np.loadtxt(f,usecols=(1,2,3))
    assert(len(positions)==natoms)
    return np.array(cell),np.array(positions)

# note: this can act on a vector of vectors
def pbc(self,cell,vector):
    return vector-np.floor(vector/cell+0.5)*cell

def write_trajectory(file,trajectory,*,wrapatoms=False):
    with open(file,"w") as f:
        for cell,positions in trajectory:
            print("%d" % len(positions), file=f)
            print("%f %f %f" % (cell[0], cell[1], cell[2]), file=f)
            if wrapatoms:
                positions = pbc(cell,positions)
            np.savetxt(f,positions,fmt="Ar %10.7f %10.7f %10.7f")


class SimpleMD:
    def __init__(self,*,
                temperature=1.0,
                maxneighbors=5000,
                tstep=0.005,
                friction=0.0,
                forcecutoff=2.5,
                listcutoff=3.0,
                nstep=1,
                nconfig=10,
                nstat=1,
                idum=0,
                wrapatoms=False,
                statfile=None,
                inputfile=None,
                outputfile=None,
                trajfile=None,
                cell=None,
                positions=None):


        self.iv=np.zeros(32,dtype=int)
        self.iy=0
        self.iset=0
        self.gset=0.0
        self.write_positions_first=True
        self.write_statistics_first=True
        self.write_statistics_last_time_reopened=0
        self.write_statistics_fp=None

        self.temperature=temperature
        self.maxneighbors=maxneighbors
        self.tstep=tstep
        self.friction=friction
        self.forcecutoff=forcecutoff
        self.listcutoff=listcutoff
        self.nstep=nstep
        self.nconfig=nconfig
        self.nstat=nstat
        self.idum=idum
        self.wrapatoms=wrapatoms
        self.statfile=statfile
        self.trajfile=trajfile
        self.outputfile=outputfile
        self.inputfile=inputfile
        self.positions=+positions
        self.cell=cell

        self.statfile_f=None

        self.statistics=[]
        self.trajectory=[]

        if positions is not None and cell is None:
           raise Exception("either pass both cell and positions or none of them")
        if positions is None and cell is not None:
           raise Exception("either pass both cell and positions or none of them")

        if self.inputfile is None and positions is None:
            raise Exception("Either specify input file or positions")
        if self.inputfile is not None and positions is not None:
            raise Exception("Either specify input file or positions")

    def randomize_velocities(self,temperature,masses,random):
       return np.sqrt(temperature/masses)[:,np.newaxis]*random.Gaussian(shape=(len(masses),3))

    def check_list(self,positions,positions0,listcutoff,forcecutoff):
       delta2=(0.5*(listcutoff-forcecutoff))*(0.5*(listcutoff-forcecutoff))
       disp2=np.sum((positions-positions0)**2,axis=1)
       return np.any(disp2>delta2)

    def compute_engkin(self,masses,velocities):
        return 0.5*np.sum(masses*np.sum(velocities**2,axis=1))

    def thermostat(self,masses,dt,friction,temperature,velocities,engint,random):
        c1=np.exp(-friction*dt)
        c2=np.sqrt((1.0-c1*c1)*temperature)/np.sqrt(masses)
        engint+=0.5*np.sum(masses*np.sum(velocities**2,axis=1))
        velocities=c1*velocities+c2[:,np.newaxis]*random.Gaussian(shape=velocities.shape)
        engint-=0.5*np.sum(masses*np.sum(velocities**2,axis=1))
        return velocities,engint

    def write_positions(self,cell,positions,wrapatoms=False):
        if self.trajfile:
            mode="w"
            if self.write_positions_first:
                self.write_positions_first = False
                mode="w"
            else:
                mode="a"
     
            with open(self.trajfile,mode) as f:
                print("%d" % len(positions), file=f)
                print("%f %f %f" % (cell[0], cell[1], cell[2]), file=f)
                if wrapatoms:
                    positions = pbc(cell,positions)
                np.savetxt(f,positions,fmt="Ar %10.7f %10.7f %10.7f")
        else:
            self.trajectory.append((cell,+positions))

    def write_final_positions(self,cell,positions,wrapatoms=False):
        if self.outputfile:
            with open(self.outputfile,"w") as f:
                print("%d" % len(positions), file=f)
                print("%f %f %f" % (cell[0], cell[1], cell[2]), file=f)
                if wrapatoms:
                    positions = pbc(cell,positions)
                np.savetxt(f,positions,fmt="Ar %10.7f %10.7f %10.7f")
        else:
            self.output=(cell,positions)

    def write_statistics(self,istep,tstep,natoms,engkin,engconf,engint):
         if self.statfile:
             if self.write_statistics_fp is None:
                 self.write_statistics_fp = open(self.statfile, "w")
             if istep-self.write_statistics_last_time_reopened>100:
                 self.write_statistics_fp.close()
                 self.write_statistics_fp = open(self.statfile, "a")
                 self.write_statistics_last_time_reopened=istep
             print("%d %f %f %f %f %f" %
                   (istep,istep*tstep,2.0*engkin/(3.0*natoms),engconf,engkin+engconf,engkin+engconf+engint),
                   file=self.write_statistics_fp)
         else:
              self.statistics.append((istep,istep*tstep,2.0*engkin/(3.0*natoms),engconf,engkin+engconf,engkin+engconf+engint))

    def run(self):
        if self.positions is None:
            self.cell,self.positions=read_positions(self.inputfile)

        random=Random(self.idum)

        # masses are hardcoded to 1
        masses=np.ones(len(self.positions))

        # energy integral initialized to 0
        engint=0.0

        # velocities are randomized according to temperature
        velocities=self.randomize_velocities(self.temperature,masses,random)

        # allocate space for neighbor lists
        nlist=np.zeros(self.maxneighbors*len(self.positions), dtype=int)
        point=np.zeros(len(self.positions)+1, dtype=int)
        # neighbour list are computed
        _compute_list(self.cell, self.positions, self.listcutoff, nlist, point)

        #print("Neighbour list recomputed at step ",0)
        #print("List size: ",len(nlist))

        # reference positions are saved
        positions0=+self.positions

        forces=np.zeros(shape=self.positions.shape)

        # forces are computed before starting md
        engconf= _compute_forces(self.cell, self.positions, self.forcecutoff, nlist, point, forces)

        # here is the main md loop
        # Langevin thermostat is applied before and after a velocity-Verlet integrator
        # the overall structure is:
        #   thermostat
        #   update velocities
        #   update positions
        #   (eventually recompute neighbour list)
        #   compute forces
        #   update velocities
        #   thermostat
        #   (eventually dump output informations)

        for istep in tqdm(range(self.nstep)):
            if self.friction>0:
                velocities,engint = self.thermostat(
                        masses,0.5*self.tstep,self.friction,self.temperature,velocities,engint,random)

            velocities+=forces*0.5*self.tstep/masses[:,np.newaxis]
            self.positions+=velocities*self.tstep

            check_list=self.check_list(self.positions,positions0,self.listcutoff,self.forcecutoff)
            if check_list:
                _compute_list(self.cell, self.positions, self.listcutoff, nlist, point)
                positions0=+self.positions
                #print("Neighbour list recomputed at step ",istep)
                #print("List size: ",len(nlist))

            engconf = _compute_forces(self.cell, self.positions, self.forcecutoff, nlist, point, forces)

            velocities+=forces*0.5*self.tstep/masses[:,np.newaxis]

            if self.friction>0.0:
                velocities,engint = self.thermostat(
                        masses,0.5*self.tstep,self.friction,self.temperature,velocities,engint,random)

            if (istep+1)%self.nconfig==0:
                self.write_positions(self.cell,self.positions,self.wrapatoms)
            if (istep+1)%self.nstat==0:
                engkin = self.compute_engkin(masses,velocities)
                self.write_statistics(istep+1,self.tstep,len(self.positions),engkin,engconf,engint)

        self.write_final_positions(self.cell,self.positions,self.wrapatoms)

        if self.write_statistics_fp is not None:
            self.write_statistics_fp.close()

if __name__ == "__main__":
    # read from stdin, store on a temporary file, and read in a dictionary
    input = sys.stdin.read()
    with tempfile.NamedTemporaryFile("w+t") as tmp:
        tmp.write(input)
        tmp.flush()
        keys=read_input(tmp.name)

    # when running from the command line, used should provide all file names
    if "inputfile" not in keys:
        raise Exception("Specify input file")
    if "outputfile" not in keys:
        raise Exception("Specify output file")
    if "trajfile" not in keys:
        raise Exception("Specify traj file")
    if "statfile" not in keys:
        raise Exception("Specify stat file")

    # run the actual simulation
    simplemd=SimpleMD(**keys)
    simplemd.run()

